File: python/linked_lists/sum_lists.py
"""
question:
You have two numbers represented by a linked list, where each node contains a single digit. The digits are stored in
reverse order, such that the 1's digit is at the head of the list. Write a function that adds the two numbers and
returns the sum as a linked list. follow up question: Suppose the digits are stored in forward order.

e.g.
Input:  (7 ->  1  ->  6)  +  (5  ->  9  ->  2) . That is, 617  +  295.
Output: 2 - > 1 - > 9. That  is,  912.

source:
McDowell, Gayle Laakmann, Cracking the Coding Interview: 189 Programming Questions and Solutions 6th Edition (2015) 214.
"""
from linked_lists.linked_list import Node
from linked_lists.linked_list import LinkedList
from linked_lists.linked_list import LinkedListWithTail

def sum_nodes(a, b, carry):
    """get sum of node values plus carry"""

    res = carry

    if a:
        res += a.value
    if b:
        res += b.value

    return res

def get_list_length(head):
    """get length of linked list"""

    cur = head
    len = 0

    while cur:
        len += 1
        cur = cur.next

    return len

def pad_smaller(a, b):
    """pad smaller list with zeros until same size"""

    a_len = get_list_length(a)
    b_len = get_list_length(b)

    if a_len > b_len:
        for i in range(a_len - b_len):
            new_node = Node(0)
            new_node.next = b
            b = new_node
    elif b_len > a_len:
        for j in range(b_len - a_len):
            new_node = Node(0)
            new_node.next = a
            a = new_node
    else:
        pass

    return (a,b)

def sum_lists_rev_order(a, b):
    """
    Solution:
    Iterate through reverse-order linked lists.
    """

    res = LinkedListWithTail()
    carry = 0

    while a or b or carry:

        # calc sum
        sum = sum_nodes(a, b, carry)

        # set digit
        res.append(sum % 10)

        # get next
        if a:
            a = a.next
        if b:
            b = b.next

        # carry the one
        carry = 1 if sum > 9 else 0

    return res.head


def sum_lists_fwd_order(a, b):
    """
    Solution:
    Recurse through forward-order linked lists.
    """

    res = LinkedList()
    a, b = pad_smaller(a, b)

    def sum_next(a, b, res):

        # base case
        if not a:
            return 0

        # set next digit
        carry = sum_next(a.next, b.next, res)

        # set this digit
        sum = sum_nodes(a, b, carry)
        res.push(sum % 10)

        # return carry
        return 1 if sum > 9 else 0

    # sum lists get carry
    carry = sum_next(a, b, res)

    # push final 1 if carry
    if carry:
        res.push(1)

    return res.head

"""
test
"""
def linked_list(values):
    ll = LinkedList()
    for v in values:
        ll.push(v)
    return ll.head

def string_list(head):
    node = head
    res = ""
    while node:
        res += '{}'.format(node.value)
        if node.next:
            res += ' -> '
        node = node.next
    return res

def run_test(fn, x, y):
    return string_list(fn(linked_list(x), linked_list(y)))

def test_sum_lists_rev_order():
    assert run_test(sum_lists_rev_order, [6,1,7], [9,5]) == '2 -> 1 -> 7'

def test_sum_list_fwd_order():
    assert run_test(sum_lists_fwd_order, [7,1,6], [5,9]) == '7 -> 1 -> 2'

File: javascript/linked_lists/sum_lists.js
/*
You are given two non-empty linked lists representing two non-negative integers. The most significant digit comes first and each of their nodes contains a single digit. Add the two numbers and return the sum as a linked list. You may assume the two numbers do not contain any leading zero, except the number 0 itself.

source: Add two numbers II (lc 445) - https://leetcode.com/problems/add-two-numbers-ii/
*/

/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {

  // if one number has fewer digits, find it.
  let [l1Len, l2Len] = [getLen(l1), getLen(l2)];
  let [sml, big] = (l1Len <= l2Len) ? [l1, l2] : [l2, l1];

  // add equal length lists (pad smaller)
  let [head, carry] = add(pad(sml, Math.abs(l1Len - l2Len)), big);

  // handle final carry
  if (carry) {
      let newNode = new ListNode(carry);
      newNode.next = head;
      head = newNode;
  }

  return head;
};

function add(n1,n2) { // postOrder dft

  if (!n1 && !n2) return [null, 0]

  // 1. add rest of digits (excluding first)
  let [head, carry] = add(n1.next,n2.next);

  // 2. add first digits + carry and append to rest
  let sum = n1.val + n2.val + carry;
  let newNode = new ListNode(sum % 10);
  newNode.next = head;
  return [newNode, Math.floor(sum/10)];
}

function getLen(head) {
  let cur = head, len = 0;
  while (cur) { len++; cur = cur.next; }
  return len;
}

function pad(list, x) {
  if (x > 0) {
      for (i = 0; i < x; i++) {
          let pad = new ListNode(0);
          pad.next = list;
          list = pad;
      }
  }
  return list;
}
